src/vm/runners/cairo_runner.rs

```diff
@@ -3458,6 +3496,28 @@ mod tests {
         assert_eq!(vm.segments.public_memory_offsets.get(&1), Some(&vec![]));
     }
 
+    #[test]
+    fn read_return_values_test() {
+        let mut program = program!();
+        program.data = vec_data![(1), (2), (3), (4), (5), (6), (7), (8)];
+        //Program data len = 8
+        let mut cairo_runner = cairo_runner!(program, "plain", true);
+        cairo_runner.program_base = Some(Relocatable::from((0, 0)));
+        cairo_runner.execution_base = Some(Relocatable::from((1, 0)));
+        cairo_runner.run_ended = true;
+        let mut vm = vm!();
+        assert_eq!(cairo_runner.finalize_segments(&mut vm), Ok(()));
+        assert!(cairo_runner.segments_finalized);
+        //Check values written by first call to segments.finalize()
+        assert_eq!(vm.segments.segment_sizes.get(&0), Some(&8_usize));
+
+        assert_eq!(cairo_runner.read_return_values(&vm), Ok(()));
+        assert_eq!(
+            cairo_runner.execution_public_memory.unwrap(),
+            Vec::<usize>::new()
+        );
+    }
```

Comment:
We also need to test error cases and cases where there's public memory after running.